m2cgen/interpreters/python/interpreter.py

```diff
@@ -27,19 +29,27 @@ def interpret(self, expr):
             last_result = self._do_interpret(expr)
             self._cg.add_return_statement(last_result)
 
-        if self.with_vectors or self.with_math_module:
+        if self.with_math_module:
+            self._cg.add_dependency("math")
+
+        if self.with_vector_operations:
             self._cg.add_dependency("numpy", alias="np")
 
         return self._cg.code
 
     def interpret_bin_vector_expr(self, expr, **kwargs):
+        self.with_vector_operations = True
         return self._cg.infix_expression(
-            left=self._do_interpret(expr.left, **kwargs),
+            left=self._cg.array_convert_to_numpy(
```

Comment:
I'm not sure I understand this change. Why can't keep using `vector_init` for this purpose?

Reply:
Because `[1, 2, 3] + [4, 5, 6] != [5, 7, 9]` in Python.

Reply:
My question is rather why do we need raw Python lists in a generated code in the first place. Why can't it always be a numpy array?

Reply:
I'm guessing that this the part where we may fallback to Python lists to eliminate the `numpy` dependency but I'd rather always have a consistent representation of the vector (especially considering that we may return it from the `score` function). So I'd argue that every time when we need vectors, it should be `numpy` arrays.

Reply:
Yeah, this is about eliminating heavy redundant dependency from runtime. I think that this is quite more important rather than have consistent return value. There are many frameworks which return either `type1`, or `type2`, depending on some conditions. BTW, `score` at present may return `double` and `double[]`, so this can be treated as inconsistency as well.

I think that from the user side it's much easier to wrap `score` function call into `np.array()` only once, if they need consistent output, rather than manually edit generated function removing dozens of `np.assarray()` entries, if they do not have/want/able to install `numpy`.

Reply:
*Joke* 😃 : It's time to change project's description: 
> Transform ML models into a native code (Java, C, Python, Go, JavaScript) with zero dependencies

Reply:
> BTW, score at present may return double and double[], so this can be treated as inconsistency as well.
Well there is a distinction between scalar and vector, but here we're introducing different flavors of vectors. 

Additionally we break a backward compatibility here, which is not great. 

I'm positive about the benefit of using `math` vs `numpy` in certain scenarios, but I'm not fully convinced about the whole vector thing.

@krinart you're a better Python expert than I am. What's your opinion?

Reply:
> Well there is a distinction between scalar and vector, but here we're introducing different flavors of vectors.
Additionally we break backward compatibility, which is not great.

Agree! I only want to say that the project has version `0.4.0`, which means that, according to the
semantic versioning, its API is subject to change.

Reply:
After discussion with @krinart I think it may not be that bad in Python world. We also discussed the possibility of dropping numpy dependency altogether in favor of manually written vector multiplication like we do in other languages. This PR can be the first step in that direction. Let's merge it. @StrikerRUS thanks a lot for your effort 👍

Reply:
@izeigerman  Thanks a lot for considering this change!
> We also discussed the possibility of dropping numpy dependency altogether in favor of manually written vector multiplication like we do in other languages. 

It's a great step towards isolated environments where it's not so straightforward to install any dependency! I guess they are the majority of the target audience of the package. As of now only vector addition and multiplication by number should be supported, it will not hurt the performance too much.

However, with the growth of supported algorithms number there can be needed more non-trivial vector operations where `numpy` speed will be crucial. So I think there can be an option which will provide users the possibility to choose either they need a runtime with zero dependencies but slow, or one with a few number of dependencies but significantly faster.

cc @krinart